ilastik/applets/dataSelection/opDataSelection.py

```diff
@@ -715,105 +717,93 @@ def __init__(
+            if datasetInfo.shape5d.c > 1:
+                meta = {"channel_names": [f"{role_name}-{i}" for i in range(datasetInfo.shape5d.c)]}
             else:
-                # No forced axisorder is supplied. Use original axisorder as
-                # output order: it is assumed by the export-applet, that the
-                # an OpReorderAxes operator is added in the beginning
-                output_order = providerSlot.meta.getAxisKeys()
-            if "c" not in output_order:
-                output_order += "c"
-
-            op5 = OpReorderAxes(parent=self, AxisOrder=output_order, Input=providerSlot)
-            self._opReaders.append(op5)
+                meta = {"channel_names": [role_name]}
+            data_provider.meta.update(meta)
 
+            output_order = self._get_output_axis_order(data_provider)
+            op5 = OpReorderAxes(parent=self, AxisOrder=output_order, Input=data_provider)
             self.Image.connect(op5.Output)
-
             self.AllowLabels.setValue(datasetInfo.allowLabels)
-
             if self.Image.meta.nickname is not None:
                 datasetInfo.nickname = self.Image.meta.nickname
-
             self.ImageName.setValue(datasetInfo.nickname)
 
         except:
-            self.internalCleanup()
+            self._clean_up_all_children()
             raise
 
+    def _get_output_axis_order(self, data_provider: OutputSlot) -> str:
+        if self.forceAxisOrder:
+            assert isinstance(self.forceAxisOrder, list), "forceAxisOrder should be a *list* of preferred axis orders"
+
+            # Before we re-order, make sure no non-singleton
+            #  axes would be dropped by the forced order.
+            tagged_provider_shape = data_provider.meta.getTaggedShape()
+            minimal_axes = {k for k, v in tagged_provider_shape.items() if v > 1}
+
+            # Pick the shortest of the possible 'forced' orders that
+            # still contains all the axes of the original dataset.
+            candidate_orders = list(self.forceAxisOrder)
+            candidate_orders = [order for order in candidate_orders if minimal_axes.issubset(set(order))]
+
+            if len(candidate_orders) == 0:
+                msg = (
+                    f"The axes of your dataset ({data_provider.meta.getAxisKeys()}) are not compatible with "
+                    f"any of the allowed axis configurations used by this workflow ({self.forceAxisOrder})."
+                )
+                raise DatasetConstraintError("DataSelection", msg)
+
+            output_order = sorted(candidate_orders, key=len)[0]  # the shortest one
+            output_order = "".join(output_order)
+        else:
+            # No forced axisorder is supplied. Use original axisorder as
+            # output order: it is assumed by the export-applet, that the
+            # an OpReorderAxes operator is added in the beginning
+            output_order = data_provider.meta.getAxisKeys()
```

Comment:
```suggestion
            output_order = "".join(data_provider.meta.getAxisKeys())
```

`getAxisKeys` returns `List[str]`

Reply:
I guess this was never a problem because the downstream OpReorderAxes does the same join again in setupOutputs... but yeah this should be consistent between the different paths.

Reply:
Entertainingly:
```
a = ["x", "y"]
a += "c" # This is fine
a = a + "c" # TypeError
```